kame/kame/mip6config/mip6config.8
.\"	$KAME: mip6config.8,v 1.3 2000/11/10 12:12:11 itojun Exp $
.\"
.\" Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Nov 22, 1999
.Dt MIP6CONFIG 8
.Os KAME
.\"
.Sh NAME
.Nm mip6config
.Nd Configure the Mobile IPv6 kernel module
.\"
.Sh SYNOPSIS
.Nm mip6config
.Bk -words
.Fl H Ar home_address/plen@if%ha_addr
.Ek
.Bk -words
.Op Fl F Ar foreign_address/plen@if
.Ek
.Bk -words
.Op Fl b Ar bu_lifetime
.Ek
.Bk -words
.Op Fl w Ar br_time
.Ek
.Bk -words
.Op Fl l Ar hr_lifetime
.Ek
.Bk -words
.Op Fl dpqrt Li 0|1
.Ek
.Bk -words
.Op Fl e Li 0|1|2
.Ek
.Nm mip6config
.Fl a
.Bk -words
.Op Fl F Ar foreign_address/plen@if
.Ek
.Bk -words
.Op Fl b Ar bu_lifetime
.Ek
.Bk -words
.Op Fl w Ar br_time
.Ek
.Bk -words
.Op Fl l Ar hr_lifetime
.Ek
.Bk -words
.Op Fl Ar dpqrt Li 0|1
.Ek
.Bk -words
.Op Fl e Li 0|1|2
.Ek
.Nm mip6config
.Fl g
.Op Fl w Ar br_time
.Op Fl y Ar ha_pref
.Op Fl l Ar hr_lifetime
.Op Fl mqu Li 0|1
.Nm mip6config
.Fl x
.Nm mip6config
.Fl f Ar filename
.Nm mip6config
.Bk -words
.Op Fl F Ar foreign_address/plen@if
.Ek
.Bk -words
.Op Fl b Ar bu_lifetime
.Ek
.Bk -words
.Op Fl w Ar br_time
.Ek
.Bk -words
.Op Fl y Ar ha_pref
.Ek
.Bk -words
.Op Fl l Ar hr_lifetime
.Ek
.Bk -words
.Op Fl dmpqrtu Li 0|1
.Ek
.Bk -words
.Op Fl e Li 0|1|2
.Ek
.Nm mip6config
.Fl E Ar address
.\"
.Sh DESCRIPTION
The Mobile IPv6 implementation has been integrated as part of the FreeBSD 
kernel.
It consists of three different parts: The correspondent node, which 
is mandatory, must always be included for an IPv6 node to claim "IPv6 
compliance" The home agent part is optional and required if a router is 
acting as a home agent at the home network.
The mobile node part is optional 
and is required by a host (router or host) to be able to move between 
different sub-networks while maintaining transport and higher layer 
connections.
A node may not act as mobile node and home agent simultaneously.
.Pp
There are two different ways of including the Mobile IPv6 software in the
kernel.
The first alternative is to add it during compilation.
The 
correspondent node part must always be included and the optional parts can
be included if desired.
The second alternative is to add the correspondent node
part at compilation time and add the optional parts as modules during 
runtime.
.Pp
The
.Nm
command is used for configuration of the Mobile IPv6 code.
Depending on which 
part of the code that has been loaded in the kernel, different options will 
be enabled/disabled.
.\"
.Ss USAGE
The
.Fl a
option is for autoconfiguration 
.\"
.Ss OPTIONS
.Bl -tag -width Ds
.It Fl E Ar address
Remove a foreign address.
.It Fl F Ar foreign_address/plen@if
Manually assign a foreign address.
.It Fl H Ar home_address/plen@if%ha_addr
Manually assign a home address and a home agent.
.It Fl a
Enable auto configuration of the home address.
.It Fl b Ar bu_lifetime
Set the binding update lifetime.
.It Fl d Li 0|1
Disable/Enable debug messages.
.It Fl e Li 0|1|2
Disable/Enable eager Movement Detection.
Two levels are provided.
At level 1 more aggressive parameter settings are used.
At level 2, in addition, new prefixes are taken as indication of handover, rather than the expiration of the previous default router.
.It Fl f Ar filename
Read configuration from file.
.It Fl g
Enable Home Agent functionality.
.It Fl h
Show a simple usage information.
.It Fl l Ar hr_lifetime
Set the home registration lifetime.
.It Fl m Li 0|1
Disable/Enable forwarding of Site local Multicast dest addresses.
.It Fl p Li 0|1
Disable/Enable link layer promiscuous mode.
.It Fl q Li 0|1
Disable/Enable sending binding requests to the mobile node.
.It Fl r Li 0|1
Disable/Enable sending binding updates to a correspondent node (Route optimisation).
.It Fl t Li 0|1
Disable/Enable tunneling of packets from the nobile node to a correspondent node via the home agent
.It Fl u Li 0|1
Disable/Enable forwarding of Site local Unicast dest addresses
.It Fl w Ar br_time
Set the binding request time.
.It Fl x
De-activate Mobile Node/Home Agent functionality (exit).
.El
.Sh EXAMPLES
This section contains some simple examples.
.Pp
To load and activate the Mobile IPv6 functionality, type
.Bd -literal -offset indent
# kldload ./mip6.ko
# mip6config <...>
.Ed
.Pp
To deactivate the module and unload it from the kernel, type
.Bd -literal -offset indent
# kldunload -n mip6
.Ed
.Sh AUTHOR
Magnus Braathen <magnus.braathen@era.ericsson.se>


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/mip6config/mip6config.c
/*
 * Copyright (C) 1995, 1996, 1997, 1998, 1999 and 2000 WIDE Project.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Copyright (c) 1999 and 2000 Ericsson Radio Systems AB
 * All rights reserved.
 *
 * Author:  Hesham Soliman <Hesham.Soliman@ericsson.com.au>
 *          Magnus Braathen <Magnus.Braathen@era.ericsson.se>
 *
 * $Id: mip6config.c,v 1.6 2000/03/18 03:05:45 itojun Exp $
 *
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <netdb.h>
#include <err.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet6/mip6_common.h>
#include <err.h>
#include "mip6config.h"

static int mip6_module;
static int startmip6;
static int s;

static char *configfilename;

extern void print_err __P((int));

static int
upd_kernel(u_long cmd, struct mip6_input_data *args)
{
    struct mip6_input_data dummy;

    if (args == NULL) {
        bzero(&dummy, sizeof(dummy));
        args = &dummy;    /* ioctl() third argument must be non-NULL */
    }

    /* Note: max transfer size is PAGE_SIZE (4096 bytes?) */
    if (ioctl(s, cmd, (caddr_t)args) < 0) {
		perror("ioctl");
        printf("Make sure kernel supports MIP6.\n");
        exit(1);
    }
    return 0;  /* XXXYYY Better error handling needed? */
}

/*
mip6config -F (Set Default foreign IP Address)
mip6config -H (Write home address)
mip6config -E (Remove default foreign address from list)

mip6config -f (Read from file)
mip6config -b (Default BU lifetime)
mip6config -w (Set time when CN should send Binding request)
mip6config -y (HA preference)
mip6config -l (Default lifetime for home registration, not BU)

mip6config -g (Enable HA functionality)
mip6config -u (Enable forwarding of Site local Unicast dest addresses)
mip6config -m (Enable forwarding of Site local Multicast dest addresses)
mip6config -p (Enable lnk layer promiscuous mode)
mip6config -r (Enable sending BU to CN, ie.Route optimisation on/off )
mip6config -t (Enable tunneling of packets from MN to CN via HA)
mip6config -q (Enable sending BR to the MN)
mip6config -d (Enable debug)
mip6config -a (Allow autoconfiguration of Home address)
mip6config -e (Enable eager Movement Detection)
mip6config -x (De-activate MN/HA functionality)
*/

static void
usage()
{
	fprintf(stderr, "\n%s\n%s\n%s%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n",
"usage: mip6config -a [-F foreign_address/plen@if] [-b bu_lifetime]",
"                     [-w br_time] [-l hr_lifetime] [-depqrt 0/1]",
"       mip6config -H home_address/plen@if%ha_addr ",
                     "[-F foreign_address/plen@if]",
"                     [-b bu_lifetime] [-w br_time] [-l hr_lifetime]",
"                     [-depqrt 0/1]",
"       mip6config -g [-w br_time] [-y ha_pref] [-l hr_lifetime] [-mqu 0/1]",
"       mip6config -x",
"       mip6config -E address",
"       mip6config -f file",
"       mip6config [-F foreign_address/plen@if] [-b bu_lifetime] [-w br_time]",
"                  [-y ha_pref] [-l hr_lifetime] [-dempqrtu 0/1]");
	exit(1);
}

/* parses a line like ipv6addr/prefixlen@interface */
static int
parse_addr(char *addrline, char **addr, int *plen, char **iface)
{
    char *plen_ptr;

    *addr = addrline;

    if ((plen_ptr = strchr(*addr, '/')) == NULL) {
        return -1;
    }

    *plen_ptr = '\0';
    plen_ptr++;

    *plen = atoi(plen_ptr);

    if ((*iface = strchr(plen_ptr, '@')) == NULL) {
        return -1;
    }

    **iface = '\0';
    (*iface)++;

    return 0;
}

/* Returns the address in network order */
static int
getaddress(char *address, struct in6_addr *in6addr)
{
    if (inet_pton(AF_INET6, address, in6addr) == NULL) {
        struct hostent *hp;
 	if ((hp = gethostbyname2(address, AF_INET6)) == NULL)
            return -1;
        else
            memcpy(in6addr, hp->h_addr_list[0], sizeof(struct in6_addr));
    }
    return 0;
}

/* Parsing functions */
static int
set_homeaddr(char *homeaddr, u_long command)
{
    struct mip6_input_data input;
    char *address, *interface, *homeagent = NULL;
    int   plen;
    int retval;

    if (parse_addr(homeaddr, &address, &plen, &interface) < 0) {
        warnx("error parsing home address %s\n", homeaddr);
        return -1;
    }

    if ((homeagent = strchr(interface, '%')) != NULL) {
        *homeagent = '\0';
        homeagent++;
    }

#ifdef DEBUG
    printf("set_homeaddr: address: %s, plen: %d, interface: %s, homeagent: %s\n",
           address, plen, interface, homeagent);
#endif /* DEBUG */

    if (getaddress(address, &input.ip6_addr) < 0) {
        warnx("unknown home address %s\n", address);
        return -1;
    }

    if (homeagent == NULL || *homeagent == '\0' || *homeagent == ' ')
        input.ha_addr = in6addr_any;
    else {
 	if (getaddress(homeagent, &input.ha_addr) < 0) {
            warnx("unknown homeagent address %s\n", homeagent);
            return -1;
        }
    }

    input.prefix_len = plen;
    strcpy(input.if_name, interface);

    mip6_module = MIP6_MN_MODULE;
    startmip6 = 1;

    if ((retval = upd_kernel(command, (void *)&input)) > 0) {
        /* error */
        print_err(retval);
    }

    return retval;
}

static int
del_coaddr(char *coaddr, u_long command)
{
    struct mip6_input_data input;
    char *address, *interface;
    int   plen;
    int retval;

    if (parse_addr(coaddr, &address, &plen, &interface) < 0) {
        warnx("error parsing c/o address %s\n", coaddr);
        return -1;
    }

#ifdef DEBUG
    printf("del_coaddr: address: %s, plen: %d, interface: %s\n",
           address, plen, interface);
#endif /* DEBUG */

    if (getaddress(address, &input.ip6_addr) < 0) {
        warnx("unknown c/o address %s\n", address);
        return -1;
    }

    input.prefix_len = plen;
    strcpy(input.if_name, interface);

    if ((retval = upd_kernel(command, (void *)&input)) > 0) {
        /* error */
        print_err(retval);
    }

    return retval;
}

static int
set_coaddr(char *coaddr, u_long command)
{
    struct mip6_input_data input;
    char *address, *interface;
    int   plen;
    int retval;

    if (parse_addr(coaddr, &address, &plen, &interface) < 0) {
        warnx("error parsing c/o address %s\n", coaddr);
        return -1;
    }

#ifdef DEBUG
    printf("set_coaddr: address: %s, plen: %d, interface: %s\n",
           address, plen, interface);
#endif /* DEBUG */

    if (getaddress(address, &input.ip6_addr) < 0) {
        warnx("unknown c/o address %s\n", address);
        return -1;
    }

    input.prefix_len = plen;
    strcpy(input.if_name, interface);

    if ((retval = upd_kernel(command, (void *)&input)) > 0) {
        /* error */
        print_err(retval);
    }

    return retval;
}

static int
set_enable(char *enable, u_long command)
{
    struct mip6_input_data input;
    int retval;

#ifdef DEBUG
    printf("set_enable: %s, command: %lu\n", enable, command);
#endif /* DEBUG */

    if ((enable == NULL) || (enable != NULL && atoi(enable) >= 1))
        input.value = 1;
    else
        input.value = 0;

    if (command == SIOCSAUTOCONFIG_MIP6) {
	    mip6_module = MIP6_MN_MODULE;
	    startmip6 = 1;
    }

    if ((retval = upd_kernel(command, (void *)&input)) > 0) {
        /* error */
        print_err(retval);
    }

    return retval;
}

static int
set_value(char *value, u_long command)
{
    struct mip6_input_data input;
    int retval;

#ifdef DEBUG
    printf("set_value: %s, command: %lu\n", value, command);
#endif /* DEBUG */

    if (value == NULL)
        return -1;

    input.value = atoi(value);

    if ((retval = upd_kernel(command, (void *)&input)) > 0) {
        /* error */
        print_err(retval);
    }

    return retval;
}

static int
set_enable_ha(char *value, u_long command)
{
#ifdef DEBUG
    printf("set_enable_ha: %s, command: %lu\n", value, command);
#endif /* DEBUG */

    mip6_module = MIP6_HA_MODULE;
    startmip6 = 1;

    return 0;
}

/* Parsing template */
struct config_tmpl config_mip6[] = {
    { "homeaddr", set_homeaddr, SIOCAHOMEADDR_MIP6},
    { "coaddr", set_coaddr, SIOCACOADDR_MIP6},
    { "enable_ha", set_enable_ha, 0},
    { "bu_lifetime", set_value, SIOCSBULIFETIME_MIP6},
    { "br_update", set_value, SIOCSBRUPDATE_MIP6},
    { "autoconfig", set_enable, SIOCSAUTOCONFIG_MIP6},
    { "ha_pref", set_value, SIOCSHAPREF_MIP6},
    { "hr_lifetime", set_value, SIOCSHRLIFETIME_MIP6},
    { "fwd_sl_unicast", set_enable, SIOCSFWDSLUNICAST_MIP6},
    { "fwd_sl_multicast", set_enable, SIOCSFWDSLMULTICAST_MIP6},
    { "prom_mode", set_enable, SIOCSPROMMODE_MIP6},
    { "bu_to_cn", set_enable, SIOCSBU2CN_MIP6},
    { "rev_tunnel", set_enable, SIOCSREVTUNNEL_MIP6},
    { "enable_br", set_enable, SIOCSENABLEBR_MIP6},
    { "debug", set_enable, SIOCSDEBUG_MIP6},
    { "eager_md", set_enable, SIOCSEAGERMD_MIP6},
    { NULL, NULL, -1 }
};

static void
mip6_init(void)
{
	struct mip6_input_data input;
	int retval;

	input.value = mip6_module;

	if ((retval = upd_kernel(SIOCSATTACH_MIP6, &input)) > 0) {
		print_err(retval);
		warnx("cannot initialize \"mip6\" module.\n");
		exit(1);
	}
}

static int
read_config()
{
    FILE *configfile;
    char buf[160];
    int retval = 0;

    if ((configfile = fopen(configfilename, "r")) == NULL) {
	perror(configfilename);
	return -1;
    }

    while(fgets(buf, sizeof(buf), configfile) != NULL) {
	char *argptr;
	int   i;

	/* Remove last CR */
 	if (buf[strlen(buf) - 1] == '\n')
	    buf[strlen(buf) - 1] = '\0';

	/* Argptr is the second parameter in the config file or NULL */
 	if ((argptr = strchr(buf, ' ')) != NULL) {
	    *argptr = '\0';
	    argptr++;
	}

	/* Ignore empty lines and comments */
 	if (*buf == '\0' || *buf == '#')
	    continue;

	/* Depending on the first paramenter, call the parsing function */
	for(i = 0; config_mip6[i].comstring != NULL; i++)
     	if (strcmp(config_mip6[i].comstring, buf) == NULL) {
		retval = config_mip6[i].parse(argptr, config_mip6[i].command);
		break;
	    }

 	if (config_mip6[i].comstring == NULL)
	    warnx("ignored unknown option %s\n", buf);

 	if (retval < 0) {
	    warnx("error in %s while parsing %s\n",
		   configfilename, config_mip6[i].comstring);
	    break;
	}
    }

    if (ferror(configfile)) {
	perror("read error");
	retval = -1;
    }

    fclose(configfile);
    return retval;
}

static char *delcoaddr, *coaddr, *homeaddr, *hr_lifetime, *fwd_sl_unicast;
static char *autoconfig, *bu_lifetime, *enable_bu_to_cn, *enable_debug;
static char *enable_ha, *ha_pref, *fwd_sl_multicast, *prom_mode, *enable_br;
static char *enable_rev_tunnel, *br_update, *eager_md, *release;

int
main(int argc, char *argv[])
{
    int ch;

    /*signal(SIGTERM, mip6_exit);*/

    if (getuid() != 0){
        warnx("permission denied\n");
        exit(1);
    }

    if ((s = socket(AF_INET6, SOCK_DGRAM, 0)) < 0)
        err(1, "socket");

    while ((ch = getopt(argc, argv, "F:H:E:f:b:w:y:l:hgxu:m:p:r:t:q:d:a:e:")) != -1)
	switch(ch) {
	case 'F':
	    coaddr = optarg;
	    break;
	case 'H':
	    homeaddr = optarg;
	    break;
	case 'l':
	    hr_lifetime = optarg;
	    break;
	case 'u':
	    fwd_sl_unicast = optarg;
	    break;
	case 'm':
	    fwd_sl_multicast = optarg;
	    break;
	case 'a':
	    autoconfig = (char *)1;
	    break;
	case 'r':
	    enable_bu_to_cn = optarg;
	    break;
	case 'd':
	    enable_debug = optarg;
	    break;
	case 'E':
	    delcoaddr = optarg;
	    break;
	case 'f':
	    configfilename = optarg;
	    break;
	case 'g':
	    enable_ha = (void *)1;
	    mip6_module = MIP6_HA_MODULE;
	    startmip6 = 1;
	    break;
	case 'x':
	    release = (void *)1;
	    break;
	case 'h':
	    usage();
	    break;
	case 'p':
	    prom_mode = optarg;
	    break;
	case 'y':
	    ha_pref = optarg;
	    break;
	case 'q':
	    enable_br = optarg;
	    break;
	case 'w':
	    br_update = optarg;
	    break;
	case 't':
	    enable_rev_tunnel = optarg;
	    break;
	case 'b':
	    bu_lifetime = optarg;
	    break;
	case 'e':
	    eager_md = optarg;
	    break;
	default:
	    usage();
	}
    argv += optind;
    argc -= optind;

    if (optind < 2)
        usage();

    /* -f is the only switch allowed */
    if (configfilename &&
       (delcoaddr || coaddr || homeaddr || hr_lifetime ||
        fwd_sl_unicast || autoconfig || bu_lifetime || enable_bu_to_cn ||
        enable_debug || fwd_sl_multicast || prom_mode || enable_br ||
        enable_rev_tunnel || br_update || ha_pref || eager_md || release))
        usage();

    /* -E is the only switch allowed */
    if (delcoaddr &&
       (coaddr || homeaddr || hr_lifetime ||
        fwd_sl_unicast || autoconfig || bu_lifetime || enable_bu_to_cn ||
        enable_debug || fwd_sl_multicast || prom_mode || enable_br ||
        enable_rev_tunnel || br_update || ha_pref || eager_md || release))
        usage();

    /* -x is the only switch allowed */
    if (release &&
       (delcoaddr || coaddr || homeaddr || hr_lifetime ||
        fwd_sl_unicast || autoconfig || bu_lifetime || enable_bu_to_cn ||
        enable_debug || fwd_sl_multicast || prom_mode || enable_br ||
        enable_rev_tunnel || br_update || ha_pref || eager_md))
        usage();

    /* Only one of -a, -H and -g allowed */
    if (autoconfig &&
       (homeaddr || enable_ha))
        usage();
    if (homeaddr &&
       (autoconfig || enable_ha))
        usage();
    if (enable_ha &&
       (homeaddr || autoconfig))
        usage();

    /* -F or -e not allowed if homeagent */
    if (enable_ha &&
       (coaddr || eager_md))
	usage();

    /* -m, -u, -y and -t only allowed if homeagent */
    if (fwd_sl_unicast &&
       (autoconfig || homeaddr))
	usage();
    if (fwd_sl_multicast &&
       (autoconfig || homeaddr))
	usage();
    if (ha_pref &&
       (autoconfig || homeaddr))
	usage();
    if (enable_rev_tunnel &&
       (autoconfig || homeaddr))
	usage();

    if (configfilename) {
	if (read_config() < 0) {
	    warnx("error reading configuration file\n");
	    exit(1);
	}
    }

    if (delcoaddr)
 	if (del_coaddr(delcoaddr, SIOCDCOADDR_MIP6) > 0)
 	    exit(1);

    if (autoconfig)
 	if (set_enable(NULL, SIOCSAUTOCONFIG_MIP6) > 0)
 	    exit(1);

    if (homeaddr)
 	if (set_homeaddr(homeaddr, SIOCAHOMEADDR_MIP6) > 0)
 	    exit(1);

#if 0
    /*XXXYYY*/
    if (enable_ha)
	if (set_enable(NULL, SIOCSENABLEHA_MIP6) > 0)
	    exit(1);
#endif

    if (release)
	if (set_enable(NULL, SIOCSRELEASE_MIP6) > 0)
	exit(1);

    if (coaddr)
	if (set_coaddr(coaddr, SIOCACOADDR_MIP6) > 0)
	    exit(1);

    if (hr_lifetime)
	if (set_value(hr_lifetime, SIOCSHRLIFETIME_MIP6) > 0)
	    exit(1);

    if (fwd_sl_unicast)
	if (set_enable(fwd_sl_unicast, SIOCSFWDSLUNICAST_MIP6) > 0)
	    exit(1);

    if (fwd_sl_multicast)
	if (set_enable(fwd_sl_multicast, SIOCSFWDSLMULTICAST_MIP6) > 0)
	    exit(1);

    if (enable_bu_to_cn)
	if (set_enable(enable_bu_to_cn, SIOCSBU2CN_MIP6) > 0)
	    exit(1);

    if (enable_debug)
	if (set_enable(enable_debug, SIOCSDEBUG_MIP6) > 0)
	    exit(1);

    if (prom_mode)
	if (set_enable(prom_mode, SIOCSPROMMODE_MIP6) > 0)
	    exit(1);

    if (ha_pref)
	if (set_value(ha_pref, SIOCSHAPREF_MIP6) > 0)
	    exit(1);

    if (enable_br)
	if (set_enable(enable_br, SIOCSENABLEBR_MIP6) > 0)
	    exit(1);

    if (br_update)
	if (set_value(br_update, SIOCSBRUPDATE_MIP6) > 0)
	    exit(1);

    if (enable_rev_tunnel)
	if (set_enable(enable_rev_tunnel, SIOCSREVTUNNEL_MIP6) > 0)
	    exit(1);

    if (bu_lifetime)
	if (set_value(bu_lifetime, SIOCSBULIFETIME_MIP6) > 0)
	    exit(1);

    if (eager_md)
	if (set_enable(eager_md, SIOCSEAGERMD_MIP6) > 0)
	    exit(1);

    if (startmip6) /* Should be last */
	mip6_init();

    return 0;
}
